### PR TITLE
release-20.2: opt: add transformation rule to convert left join to inner join

### PR DESCRIPTION
**Pull request by rytaft**
Backport 1/1 commits from #53976.

/cc @cockroachdb/release

---

Release justification: bug fixes and low-risk updates to new functionality

This commit adds an exploration rule `ConvertLeftToInnerJoin`, which converts
a left join to an inner join with the same `ON` condition, and then wraps the
expression in another left join with the original left side. In order to
avoid computing the left side of the join twice, we create a `With` expression
for the left side, and then reference it with two `WithScans`. For example
(assuming `x` is the primary key of `a`):
```
  SELECT a.x, b.y FROM a LEFT JOIN b ON ST_Intersects(a.geom, b.geom);
```
is converted to:
```
  WITH a_buf AS (
    SELECT * FROM a
  )
  SELECT a_buf.x, inr.y FROM a_buf LEFT JOIN (
    SELECT * FROM a_buf JOIN b ON ST_Intersects(a_buf.geom, b.geom)
  ) AS inr
  ON a_buf.x = inr.x;
```
Note that this transformation is not desirable in the general case, but it
is useful if there is a possibility of creating an inverted join (such as in
the above example). For this reason, we only perform this transformation if
it is possible to generate an inverted join.

This transformation allows us to index-accelerate spatial left joins, which
was not possible before.

Informs #53576

Release note (performance improvement): left outer spatial joins can now
be index-accelerated, which can lead to performance improvements in some
cases.

### CODE REVIEW

**Comment by rytaft** [comment]
I added this to https://github.com/cockroachdb/cockroach/issues/53662 for approval

**Review by RaduBerinde — APPROVED** [review]
LGTM (pending approval from release team)

**Pull request merged**
